src/Features/Core/Portable/Completion/CompletionItem.cs

```diff
@@ -12,21 +12,23 @@
 namespace Microsoft.CodeAnalysis.Completion
 {
     [DebuggerDisplay("{DisplayText}")]
-    internal class CompletionItem : IComparable<CompletionItem>
+    internal sealed class CompletionItem : IComparable<CompletionItem>
     {
-        internal AsyncLazy<ImmutableArray<SymbolDisplayPart>> LazyDescription;
-
-        internal bool HasAsyncDescription { get; }
+        /// <summary>
+        /// The kind of the completion item.
+        /// This is used by the <see cref="CompletionService"/> to perform operations.
+        /// </summary>
+        public string Kind { get; }
 
         /// <summary>
-        /// An appropriate icon to present to the user for this completion item.
+        /// Miscellaneous information attached to a completion item.
         /// </summary>
-        public Glyph? Glyph { get; }
+        public ImmutableDictionary<string, string> Properties { get; }
 
         /// <summary>
-        /// The ICompletionProvider that this CompletionItem was created from.
+        /// An appropriate icon to present to the user for this completion item.
         /// </summary>
-        public virtual CompletionListProvider CompletionProvider { get; }
+        public Glyph? Glyph { get; }
```

Comment:
i agree that this should be an immutable set of tags.  note: i do worry about what the impact of that might be perfwise if we then have to map 100,000 of those to glyphs.  But maybe that can be lazily done?